Small refactorings giving hooks for Scala.js

sjrd opened the pull request:
I have reviewed the ugliest parts of Scala.js, where I need to circumvent the lack of appropriate hooks in scalac, and I came up with these three small refactorings.

Each one of these has its corresponding simplifying commit in Scala.js, in the following comparison:
https://github.com/sjrd/scala-js/compare/scala-2.11...topic/what-if-scalac-was-nicer-to-us
(in the same order).

There is no test accompanying this pull request, because there is absolutely no functionality change (it's only refactoring) and I don't quite see how to test that with partest. I'll be glad to add tests if someone can hint at me how.

(In case you care:) Incidentally, these hooks are the only ones on which Scala.js relies, in addition to being able to manipulate the phase set. Which means that _if_ they were (experimentally) overridable by plugins, then Scala.js could be implemented as a plugin, which would bring lots of advantages in terms of integration with our various tools (sbt and IDE mainly).

(There are still two files with a non-negligible amount of code duplication in Scala.js, [JSSymbolLoaders.scala](https://github.com/sjrd/scala-js/blob/topic/what-if-scalac-was-nicer-to-us/compiler/src/main/scala/scala/tools/nsc/scalajs/JSSymbolLoaders.scala) and [JSTypefileParser.scala](https://github.com/sjrd/scala-js/blob/topic/what-if-scalac-was-nicer-to-us/compiler/src/main/scala/scala/tools/nsc/scalajs/JSTypefileParser.scala), but a proper fix to these two would require much bigger refactorings in scalac; and I don't care so much about these because they are not specifically the hooks I am relying on.)

sjrd committed:
Add back the newClassLoader hook in SymbolLoaders.

This was previously an overridable method of Platform.
The loader creation was moved in
afbee09c8e0e7b1a4da1f8517c723dad9f1adb6f directly in
SymbolLoaders, but inside a method doing more logic,
namely initializeFromClassPath().

This commit simply moves the actual creation of the
class loader (`new ClassfileLoader(bin)`) into its
own method `newClassLoader`, but in SymbolLoaders.
This allows to override only that method in subclasses
of SymbolLoaders.

sjrd committed:
Move logic checking valid names from ClassPath to ClassPathContext

Other methods taking the same kind of decisions were already
in ClassPathContext, e.g., isValidName() or, in some sense,
even toBinaryName().

This makes ClassPath itself be completely agnostic of how
particular kinds of files or directories are named.

It also allows to override this logic at the context level.
Without it, overriding this logic required a fair amount of
code duplication from ClassPath.

sjrd commented:
Not sure who to ping for review, so guessing for @gkossakowski or @retronym.

xeno-by commented on src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala:
```diff
@@ -164,10 +164,17 @@ abstract class SymbolLoaders {
+   *  This is intented as a hook allowing to support loading symbols from
+   *  files other than .class files.
+   */
+  def newClassLoader(bin: AbstractFile): SymbolLoader =
```
`ClassLoader` doesn't look like a good name for something that's not actually a classloader :)

sjrd replied:
That's why it's not called `ClassfileLoader`. It might not load a classfile, but it'll still be a class. Well, it might also be an object, or a trait, or a package object...
Do we have a name for this set of concept? I'm open to suggestions for a better name for this function, I just reused the name it had when it was defined in `Platform`, since it did exactly the same thing.

gkossakowski commented:
The d415915 and 3a71db7 look almost ok.

However, the 8490128 needs more work:
- [ ] explain in the commit message why one would like to override that method and give specific example of what Scala.JS does
- [ ] if we are moving the logic responsible for context creation then let's deprecate the constructor of PathResolver which contains the same logic
- [ ] you are in dangerous territory where context is dependent on whether the inliner is enabled or not (sic!). We should really clear up why this is needed or get rid of it (preferably)

Also, how does this conflict with https://github.com/gkossakowski/scala/commits/classpath-experiments ?
In particular, I'd like to clean up my branch and submit it as PR. The end goal would be: deprecate the whole old classpath business and switch to an interface that is much more simple and gives opportunity for more efficient classpath implementation.

sjrd commented:
> explain in the commit message why one would like to override that method and give specific example of what Scala.JS does

OK.

> if we are moving the logic responsible for context creation then let's deprecate the constructor of PathResolver which contains the same logic

Yes, good point.

Should I add commits to this pull request? or rebase and push --force?

> you are in dangerous territory where context is dependent on whether the inliner is enabled or not (sic!). We should really clear up why this is needed or get rid of it (preferably)

I agree, but this does not come from my changes, it's always been there. It looked very suspicious to me too, but I don't have the knowledge to investigate why it's needed and how to get rid of it.

> Also, how does this conflict with https://github.com/gkossakowski/scala/commits/classpath-experiments ?
> In particular, I'd like to clean up my branch and submit it as PR. The end goal would be: deprecate the whole old classpath business and switch to an interface that is much more simple and gives opportunity for more efficient classpath implementation.

It does conflict in the actual implementation details, obviously. But it does not conflict with the aim of the change: all I really need is to be able to tell the classpath to consider `.jstype` files in addition to `.class` files. So if you don't hardcode the constant `".class"` in the new classpath implem, and that it can be extended from the outside to tell: "I want `.class` _and_ `.jstype`" (not _or_), then I can adapt Scala.js to use whatever hook you give me to do it.

Actually, your remark makes me look at the issue from another angle. It now seems to me that it's definitely not the classpath's responsibility to decide which files to accept; but neither is it the context's responsibility. It should actually be the _SymbolLoaders_ responsibility to decide which files to accept, since it is also his reponsibility to know _how_ to load symbols from them. So it would make sense, it seems, that the `validClassFile()` test ultimately delegates to `global.loaders.something()`, wouldn't it? (and possibly `validPackage()` and `validSourceFile()` as well) It would also solve the potential conflict at its root.

sjrd commented:
> you are in dangerous territory where context is dependent on whether the inliner is enabled or not (sic!). We should really clear up why this is needed or get rid of it (preferably)

@gkossakowski It looks like it's not necessary. I have all tests passing with a branch where I removed the logic entirely (always accepting `$class.class` files, even when inliner is off):
https://github.com/sjrd/scala/compare/topic;remove-classpath-logic-dependent-on-inline
Should I make another pull request with this change? Or does it look like too risky despite all tests passing?

gkossakowski commented:
> Should I add commits to this pull request? or rebase and push --force?

Yes, or push new commit. Either way is fine in this case because diffs are really small.

gkossakowski commented:
> It does conflict in the actual implementation details, obviously. But it does not conflict with the aim of the change: all I really need is to be able to tell the classpath to consider .jstype files in addition to .class files. So if you don't hardcode the constant ".class" in the new classpath implem, and that it can be extended from the outside to tell: "I want .class and .jstype" (not or), then I can adapt Scala.js to use whatever hook you give me to do it.

Excellent. I'll keep that in mind when working on flat classpath.

> Actually, your remark makes me look at the issue from another angle. It now seems to me that it's definitely not the classpath's responsibility to decide which files to accept; but neither is it the context's responsibility. It should actually be the SymbolLoaders responsibility to decide which files to accept, since it is also his reponsibility to know how to load symbols from them. So it would make sense, it seems, that the validClassFile() test ultimately delegates to global.loaders.something(), wouldn't it? (and possibly validPackage() and validSourceFile() as well) It would also solve the potential conflict at its root.

Good point. Actually, javac implements classpath and symbol loading the way you propose and I agree it makes sense for us to follow the same idea. I'll make sure to enable it in flat classpath.

gkossakowski commented:
> @gkossakowski It looks like it's not necessary. I have all tests passing with a branch where I removed the logic entirely (always accepting $class.class files, even when inliner is off):
> https://github.com/sjrd/scala/compare/topic;remove-classpath-logic-dependent-on-inline
> Should I make another pull request with this change? Or does it look like too risky despite all tests passing?

How about never loading '$class' files? You won't break anything in inliner because inlining of methods in traits do not work anyway, see:

https://issues.scala-lang.org/browse/SI-4767

Yes, please submit it as a separate PR. I'm willing to look into it and take the risk of regression. This logic is really strange and I'd like to get rid of. I won't be able to review it until 1st of October, though.

sjrd commented:
I fixed the first commit so that `SymbolLoaders.newClassLoader` is protected, and removed the third commit, as it is now dependent on what happens to #2977.

> Excellent. I'll keep that in mind when working on flat classpath.

Thanks

> Good point. Actually, javac implements classpath and symbol loading the way you propose and I agree it makes sense for us to follow the same idea. I'll make sure to enable it in flat classpath.

That would be great.

sjrd commented:
Ah ... The test failures must be spurious, because I reran the tests on my machine before pushing. But since I'm actually new to the build bot, and apparently it's doing weird things at the moment, I'm afraid to touch anything.
@retronym since apparently you're busy trying to decrypt kitteh's behavior, can you make a detour by this PR, please?

adriaanm commented:
Excellent! Very happy to see the discussion is leading to a nice and small set of hook methods for scala.js-as-a-plugin as well as bug fixes :-)

adriaanm commented:
The changes LGTM from a binary compatibility for partest point of view.

adriaanm commented:
As far as I can tell, @gkossakowski's comments were implemented, so I'll merge to clear out the M6 queue.

sjrd commented:
Cool, my first pull request merged into Scala :)
Thanks.

retronym commented:
Thank you! Keep them coming!

adriaanm commented:
Yes, good work! Looking forward to plugin.js :-)
